### ISSUE TITLE
[BUG] basic auth does not enable unless `enableGuestAccess: true`

### ISSUE THREAD

**Issue by ceramey1997** [Lissy93/dashy#990]
### Environment

Self-Hosted (Docker)

### Version

2.1.1

## Problem
Dashy basic authentication does not enable when adding the block below to the `conf.yml`
```
appConfig:
    auth:
        users:
        - user: myadmin
          hash: <sha256hash>
```
To make basic auth function with dashy you have to add the `enableGuestAccess: true`. If guest Access is false then basic auth functions properly.

### Please tick the boxes

- [X] You have explained the issue clearly, and included all relevant info
- [X] You are using a [supported](https://github.com/Lissy93/dashy/blob/master/.github/SECURITY.md#supported-versions) version of Dashy
- [X] You've checked that this [issue hasn't already been raised](https://github.com/Lissy93/dashy/issues?q=is%3Aissue)
- [X] You've checked the [docs](https://github.com/Lissy93/dashy/tree/master/docs#readme) and [troubleshooting](https://github.com/Lissy93/dashy/blob/master/docs/troubleshooting.md#troubleshooting) guide ![important](https://img.shields.io/badge/Important!-F6094E)
- [X] You agree to the [code of conduct](https://github.com/Lissy93/dashy/blob/master/.github/CODE_OF_CONDUCT.md#contributor-covenant-code-of-conduct)

**Comment by 15064187978** [comment]
Hello, can you help me?

**Comment by 15064187978** [comment]
https://github.com/Lissy93/dashy/issues/989

**Comment by CristianT** [comment]
I can confirm that behavior, but in the docker image 'lissy93/dashy:2.1.0' this does not happen.

**Comment by 15064187978** [comment]
> 我可以确认该行为，但在 docker 图像 'lissy93/dashy:2.1.0' 中不会发生这种情况。


I upgraded to version 2.1.1 and solved this problem

**Comment by Keskejefaislamoi** [comment]
can you copy/past your config file ? I'm under V-2.1.1, basic auth seems not working for me
guest user can edit frontend and no auth window prompted :(

`pageInfo:
  title: network.local
  description: network.local Front Dashboard
  navLinks:
    - title: GitHub
      path: https://github.com/Lissy93/dashy
    - title: Documentation
      path: https://dashy.to/docs
appConfig:
  theme: one-dark
  layout: auto
  iconSize: medium
  language: fr
  auth:
    enableGuestAccess: false
    users:
    - user: myuser
      hash: SHA256EXAMPLESTRINGFORPASSWORD
      type: admin
sections:
  - name: Getting Started
    icon: fas fa-rocket
    items:
      - title: Dashy Live
        description: Development a project management links for Dashy
        icon: https://i.ibb.co/qWWpD0v/astro-dab-128.png
        url: https://live.dashy.to/
        target: newtab
        id: 0_1481_dashylive
      - title: GitHub
        description: Source Code, Issues and Pull Requests
        url: https://github.com/lissy93/dashy
        icon: favicon
        id: 1_1481_github
      - title: Docs
        description: Configuring & Usage Documentation
        provider: Dashy.to
        icon: far fa-book
        url: https://dashy.to/docs
        id: 2_1481_docs
      - title: Showcase
        description: See how others are using Dashy
        url: https://github.com/Lissy93/dashy/blob/master/docs/showcase.md
        icon: far fa-grin-hearts
        id: 3_1481_showcase
      - title: Config Guide
        description: See full list of configuration options
        url: https://github.com/Lissy93/dashy/blob/master/docs/configuring.md
        icon: fas fa-wrench
        id: 4_1481_configguide
      - title: Support
        description: Get help with Dashy, raise a bug, or get in contact
        url: https://github.com/Lissy93/dashy/blob/master/.github/SUPPORT.md
        icon: far fa-hands-helping
        id: 5_1481_support
`

**Comment by top-cg** [comment]
Hello,
I have the same issue ... and this is really bad ... 
To have the login working, i have to set the variable : enableGuestAccess: true
The login screen is coming back ... as a guest i'm not able to see anything on the dashboard (icons) ... but i'm able to drive though the setup file.
![image](https://user-images.githubusercontent.com/18399003/207241197-d4d81a42-7f75-4adf-8c2a-052b30c35f68.png)
Very bad situation :(

**Comment by rubenmate** [comment]
I have a fresh install with this problem. Login is never shown if you don't enable guest access.

**Comment by top-cg** [comment]
> Also anyone with guest access can copy the password hashes of all users which immediately grants the guest full access to any account, including admin accounts. Dashy is 100% compromised at this point.

I agree, the guest can see the config file.

**Comment by top-cg** [comment]
Thanks @mitchnemirov ... that's a good mitigation solution.
I was curious to understand where you found this info ... and it also helped me figuring out the up-to-date documentation is located on git and not on the website !

**Comment by top-cg** [comment]
+1

**Comment by top-cg** [comment]
Up

**Comment by hubortje** [comment]
I'm facing the same issue. 
While the [solution](https://github.com/Lissy93/dashy/issues/990#issuecomment-1379719674) https://github.com/Lissy93/dashy/issues/990#issuecomment-1379719674 proposed does work for the moment, I'd be better if the login prompt appears, without guest access being enabled.

**Comment by TheRealGramdalf** [comment]
See https://github.com/Lissy93/dashy/issues/1290#issuecomment-1884249018 for a full explanation/semi solution

For notifications:
@ceramey1997 @15064187978 @CristianT @Keskejefaislamoi @top-cg @rubenmate @gemorgan @mitchnemirov @techsolo12 @hubortje 
Sorry if you didn't want to be pinged, but I figure you probably did - or you wouldn't have participated in the issue

**Comment by CrazyWolf13** [comment]
Hi
I have tested in it in the new release and can confirm I don't experience this problem anymore.
This problems seems to be related to dashy not automatically rebuilding on changes, see #1479 and https://github.com/Lissy93/dashy/issues/1290#issuecomment-1884249018 
A new release has just been published, which should fix this issue, so please upgrade to the latest version.
If this is not the case just respond back to this ticket and I'll reopen this issue.
Thanks and have a nice day.

**Comment by LonginusL33T** [comment]
I still get this problem in the new release even if I manually click rebuild. So far I've never seen the login screen.

**Comment by top-cg** [comment]
mmm ... it fails to rebuild the app :(

**Comment by CrazyWolf13** [comment]
> I still get this problem in the new release even if I manually click rebuild. So far I've never seen the login screen.

Dashy should now automatically rebuild, can you share your conf.yml? But of course remove or change your password hash.

**Comment by CrazyWolf13** [comment]
> mmm ... it fails to rebuild the app :(

@top-cg

Can you be more exact? just with it "fails" we really cannot help you.

Please share the log, browser dev console and describe what's the error.
If dashy is not working at all and it has nothing to do with auth, please create a new issue: https://github.com/Lissy93/dashy/issues/new/choose

**Comment by top-cg** [comment]
Sorry ... The report was indeed far to be derailed.
I will try to take for time asap to document with print screen

**Comment by artemdanielov** [comment]
Hello! So basic auth still does not work. After adding nessesary lines to config.yaml, dashy doesnt ask for auth and also not letting me safe config to file because i am not logged in as admin.
``appConfig:
  theme: colorful
  layout: auto
  iconSize: large
  language: en
  auth:
    users:
      - user: xxxxxx
        hash: yyyyy
        type: admin
pageInfo:
  title: Dashy``

**Comment by CrazyWolf13** [comment]
> Sorry ... The report was indeed far to be derailed.
> I will try to take for time asap to document with print screen

Can you describe how you "enable" auth, so it does not work?

Because for me it does indeed work.

**Comment by CrazyWolf13** [comment]
Okay I think I partially solved that whole auth mistery. (@Lissy93 FYI)
There were mutiple problems which were understood to be a single issue. 

Docker was not automatically rebuilding the app-> even tough the config was saved, it was not built and applied --> FIXED

------

Adding a user via GUI:
When adding user with hash and then hitting the save below the application config, then trying to click save to disk, the button is just greyed out and it is not possible to click it, therefore the auth is never saved nor applied.

------

Adding the following part below appConfig enables auth for a user test with password test and type admin.
This has been tested from me and Lissy on mutiple different systems and is confirmed to work.

appConfig:

```
  auth:
    users:
    - user: test
      hash: 9F86D081884C7D659A2FEAA0C55AD015A3BF4F1B2B0B822CD15D6C15B0F00A08
      type: admin
```

@artemdanielov Can you confirm you have pulled the latest image with `docker pull lissy93/dashy:latest` and are running version 2.1.2 ? Could you share your docker log? `sudo docker logs dashy` replace dashy with the name of the dashy container, you can see the name with `sudo docker ps`. After pulling the new image make sure to redeploy the container.

This could also be done very easily, however this will update all your docker containers: 
`sudo docker run -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower --run-once`


@LonginusL33T 
How are you activating auth? Through the GUI or through then conf.yml? Are you running the latest version of dashy? Look above for how to check and update.

**Comment by Lissy93** [comment]
Does this work if you manually edit `conf.yml`?
And like TheRealGramdalf said, clearing browser data may help too, as if you're logged in, that state will be saved locally (alternatively, just try in Incognito mode)

**Comment by LonginusL33T** [comment]
> Okay I think I partially solved that whole auth mistery. (@Lissy93 FYI) There were mutiple problems which were understood to be a single issue.
> 
> Docker was not automatically rebuilding the app-> even tough the config was saved, it was not built and applied --> FIXED
> 
> Adding a user via GUI: When adding user with hash and then hitting the save below the application config, then trying to click save to disk, the button is just greyed out and it is not possible to click it, therefore the auth is never saved nor applied.
> 
> Adding the following part below appConfig enables auth for a user test with password test and type admin. This has been tested from me and Lissy on mutiple different systems and is confirmed to work.
> 
> appConfig:
> 
> ```
>   auth:
>     users:
>     - user: test
>       hash: 9F86D081884C7D659A2FEAA0C55AD015A3BF4F1B2B0B822CD15D6C15B0F00A08
>       type: admin
> ```
> 
> @artemdanielov Can you confirm you have pulled the latest image with `docker pull lissy93/dashy:latest` and are running version 2.1.2 ? Could you share your docker log? `sudo docker logs dashy` replace dashy with the name of the dashy container, you can see the name with `sudo docker ps`. After pulling the new image make sure to redeploy the container.
> 
> This could also be done very easily, however this will update all your docker containers: `sudo docker run -v /var/run/docker.sock:/var/run/docker.sock containrrr/watchtower --run-once`
> 
> @LonginusL33T How are you activating auth? Through the GUI or through then conf.yml? Are you running the latest version of dashy? Look above for how to check and update.

![image](https://github.com/Lissy93/dashy/assets/23091103/3343772c-4793-4768-a9e5-c1411c29f789)
![image](https://github.com/Lissy93/dashy/assets/23091103/6e6f26c1-dbe8-4ed6-8dc6-1407baf3eb30)
When I save the changes through the GUI, the users array becomes empty whether I rebuild it manually or not

**Comment by artemdanielov** [comment]
opedning it in another browser or in incognito mode does not help. still - dashy is not asking for auth, but displaying notice at config save: "You cannot write changes to disk because you are not logged in as an administrator"

**Comment by TheRealGramdalf** [comment]
> opedning it in another browser or in incognito mode does not help. still - dashy is not asking for auth, but displaying notice at config save: "You cannot write changes to disk because you are not logged in as an administrator"

I had this as well; try fixing any schema errors (shown in the browser console, typically `ctrl + shift + i`, or the dashy logs) which should trigger a rebuild. It should work after that.

On a more technical note, I believe the issue may be partly with rebuilding on startup - I'm not sure if it triggers properly or not (I'd have to double check with my instance).
Regardless of whether that works fine or not, I think it would be a good idea to add the option to have a "build cache" of sorts - it should be as simple as a bind mount/volume in `/var/dist/public` (or whatever the exact path is) - this currently fails because upon a rebuild, dashy attempts to `rm -rf /var/dist/public` - which fails, since `/var/dist/public` is a bind mount (and cannot be deleted, only unmounted). Changing this to `rm -rf /var/dist/public/*` should be all that's needed, though a mechanism to detect if the config has changed while offline (and trigger accordingly) would make sense in my mind.

**Comment by TheRealGramdalf** [comment]
> doesnt look like i have any relevant errors

Try changing anything in the `conf.yml` - could be as simple as adding a trailing `/` to a URL. Just something that triggers an automatic rebuild (not a manually triggered one)

**Comment by artemdanielov** [comment]
did it - same issue. i also rebuild the container - same, it does not asks for auth(
also, for some reason, it shows this in consolу (cant check for updates):
![image](https://github.com/Lissy93/dashy/assets/59414351/0f45701e-2030-45a5-8371-528dd649e5d0)

**Comment by artemdanielov** [comment]
After enabling "guest access" the login button appears and i can log in as admin. 
So it looks like there is definatle some issues here, because i did everything suggested and nothing seems to solve the problem.
![login](https://github.com/Lissy93/dashy/assets/59414351/97987e14-3062-41b5-84e7-9d33ce572b7c)

**Comment by top-cg** [comment]
Hello,
I was able to change my setup and uncheck the option to provide guest access ... and it worked. 
![image](https://github.com/Lissy93/dashy/assets/18399003/bf65dd3e-c19c-41cd-b679-0546a767342c)
I was confuse at the begining because once this parameter was updated, asking the app to rebuild was ending in a fail.
I suppose the app rebuilt herself in the background ... and it worked.

**Comment by tianheg** [comment]
Hope this would help(https://github.com/tianheg/dashy/commit/57f96e3dabcbb4f317defe0834d15b168b410b3c), modified some files to make it work

**Comment by Lissy93** [comment]
This should now be fixed  in #1542 and released in [3.0.1](https://github.com/Lissy93/dashy/releases/tag/3.0.1)
